text/0000-inline-asm.md

```diff
@@ -0,0 +1,1099 @@
+
+The `cpuid` example above would look like this in GCC-sytle inline assembly:
+
+```C
+// GCC doesn't allow directly clobbering an input, we need
+// to use a dummy output instead.
+int ebx, ecx, discard;
+asm (
+    "cpuid"
+    : "=a"(discard), "=b"(ebx), "=c"(ecx) // outputs
+    : "a"(4), "c"(0) // inputs
+    : "edx" // clobbers
+);
+printf("L1 Cache: %i\n", ((ebx >> 22) + 1)
+    * (((ebx >> 12) & 0x3ff) + 1)
+    * ((ebx & 0xfff) + 1)
+    * (ecx + 1));
+```
+
+## Limited set of operand types
+
+The proposed set of operand types is much smaller than that which is available through GCC-style inline assembly. In particular, the proposed syntax does not include any form of memory operands and is missing many register classes.
+
+We chose to keep operand constraints as simple as possible, and in particular memory operands introduce a lot of complexity since different instruction support different addressing modes. At the same time, the exact rules for memory operands are not very well known (you are only allowed to access the data directly pointed to by the constraint) and are often gotten wrong.
+
+If we discover that there is a demand for a new register class or special operand type, we can always add it later.
+
+## Difficulty of support
+
+Inline assembly is a difficult feature to implement in a compiler backend. While LLVM does support it, this may not be the case for alternative backends such as [Cranelift][cranelift] (see [this issue][cranelift-asm]). We provide a fallback implementation using an external assembler for such backends.
+
+[cranelift]: https://cranelift.readthedocs.io/
+[cranelift-asm]: https://github.com/bytecodealliance/cranelift/issues/444
+
+## Use of double braces in the template string
+
+Because `{}` are used to denote operand placeholders in the template string, actual uses of braces in the assembly code need to be escaped with `{{` and `}}`. This is needed for AVX-512 mask registers and ARM register lists.
+
+## Post-monomorphization errors
+
+Since the code generated by `asm!` is only evaluated late in the compiler back-end, errors in the assembly code (e.g. invalid syntax, unrecognized instruction, etc) are reported during code generation unlike every other error generated by rustc. In particular this means that:
+- Since `cargo check` skips code generation, assembly code is not checked for errors.
+- `asm!` blocks that are determined to be unreachable are not checked for errors. This can even vary depending on the optimization level since inlining provides more opportunities for constant propagation.
+
+However there is a precedent in Rust for post-monomorphization errors: linker errors. Code which references a non-existent `extern` symbol will only cause an error at link-time, and this can also vary with optimization levels as dead code elimination may removed the reference to the symbol before it reaches the linker.
```

Comment:
Sorry if this was discussed elsewhere and I missed it: this seems like a potential hole in our backwards-compatibility story. I'm imagining that we aren't providing any guarantees that inline assembly *is* going to be optimized away, and so any breakage that is caused when a chunk of broken inline asm stops being optimized away should be allowed under our backcompat guarantees, yes?

Reply:
@cramertj Right.